### PR TITLE
[IMP] core: allow replacing COALESCE with UNION in sql quieries

### PR DESCRIPTION
**Pull request by yelizariev**
UNION ALL is usually better than single SELECT with COALESCE, because it allows
postgres to make a better query plan.

New feature is applied for translation field, but could be used in other cases
too.

---

task-2081307

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Yenthe666** [comment]
Hmm interesting. Any benchmarks/perf tests?

File: odoo/models.py [hunk 1]
```diff
@@ -4283,22 +4283,27 @@ def _apply_ir_rules(self, query, mode='read'):
...
     @api.model
-    def _generate_translated_field(self, table_alias, field, query):
+    def _generate_translated_field(self, table_alias, field, query, wrapper=None):
         """
         Add possibly missing JOIN with translations table to ``query`` and
         generate the expression for the translated field.
 
         :return: the qualified field name (or expression) to use for ``field``
         """
+        if not wrapper:
```
Comment (MiquelRForgeFlow) [inline odoo/models.py:4293]:
```suggestion
        if wrapper is None:
```

**Comment by yelizariev** [comment]
> Hmm interesting. Any benchmarks/perf tests?

I tested with search product.template by name:

For small databases it doesn't make it worse.:

before: 3,5 ms
after: 1,9 ms

For a database populated with dummy values (700 K products, 1.6 M products) and running on postgres v12:

before: 700 ms
after: 70 ms

With real usage the improvement should give greater speed up, because of bigger entropy, using postgres v10 (like in odoo.sh :stuck_out_tongue_winking_eye: ) and the fact that name_search for products makes several requests on translated fields

**Comment by yelizariev** [comment]
It kind of works, but does't give a speed up, but make it worse in /shop :cry: 

I'll try later with bigger database and after /shop improvements in https://github.com/odoo/odoo/pull/68113

# Performance test

* Odoo 15 (master)
* Postgres 12.5
* Demo database with website_sale installed
* 87K product.templates with 4 variants in each
* English and French language active
* Egnlish is current language
* Numbers in the table are
 * Number of queries
 * Query time, sec
 * Remaining time, sec

```
|                       | product.product search by name (few results) | product.template search by name (few results) | /shop without search | /shop with search (few results) | /shop with search (99% in results) |   |   |
|-----------------------+----------------------------------------------+-----------------------------------------------+----------------------+---------------------------------+------------------------------------+---+---|
| Before                | 12 0.340 0.040                               | 187 0.180 0.097                               | 245 0.341 0.179      | 259 0.420 0.179                 | 259 0.738 0.201                    |   |   |
| After (MAX_UNIONS=16) | 12 3.540 0.038                               | 187 0.211 0.106                               | 243 0.354 0.181      | 259 2.411 0.163                 | 259 2.810 0.227                    |   |   |
| After (MAX_UNIONS=2)  | 12 0.235 0.044                               | 187 0.205 0.109                               | 245 0.339 0.193      | 259 0.621 0.179                 | 259 1.152 0.245                    |   |   |
```

**Comment by yelizariev** [comment]
Too complicated, quite obsolete (v16+ doesn't have ir_translation table)

**Pull request closed without merging**